### PR TITLE
Added Pomodoro and Gamification Tests

### PR DESCRIPTION
**Pull request by alliekameda**
## **Pull Request Title**  
_CI tests have been added to ensure the Pomodoro and Gamification models are working properly._

---

### **What does this PR do?**  
_The included cases test default values, handling of invalid negative times, preventing duplicate Pet Collection entries, and more._  

---

### **What has been added or changed?**  
✔ **PomodoroSession Tests**: _Tests default values and negative duration._  
✔ **Pet Catalog Tests**: _Tests catalog creation and invalid drop rate._  
✔ **Pet Collection Tests**: _Tests collection creation, status toggle, and duplicate handling._  

---

### **What problem does this solve?**  
_CI testing is required to show that the models are working properly and that further work can be done on the Pomodoro app without bugs._  

---

## **Related Issue**  
🔗 This PR resolves Issue #65 

---

## **Implementation Details**  
- `IntegrityError` import is used to catch duplicate Pet Collections.

---

## **Design Documentation (If Applicable)**  

- [UML Diagram](https://drive.google.com/file/d/1xrgO12CFhIuQuG7R9_UsYOoXX24B8RNH/view)

---

### **Step-by-step testing guide**  
1️⃣ Navigate to `backend`
2️⃣ Run ` python3 manage.py test apps.submissions` or `coverage run manage.py test apps.submissions` to test the assertions
3️⃣ Run `coverage report` to see coverage percentages

---

## **Files Added/Modified**  
📂 **List of important files changed in this PR**  

| **File Name**      | **Description** |
|--------------------|----------------|
| `apps/pomodoro/tests.py`    | CI Tests |

---

## **AI Code Usage**  
🤖 **Was AI-generated code used in this PR?**  

- [x] Yes, it was use to develop tests for edge cases.

---

## **Checklist for Reviewers**  
**Things to check before approving this PR:**  
- [x] Do all the tests pass?  
- [x] Does the output of `coverage report` seem appropriate?  
- [x] Is the style and documentation appropriate?  

---

## **Next Steps & Future Enhancements**  
- **Custom pet names** _Would it be nice if students could rename their pets from the default?_  
- **Timer functionality** _Should there be separate fields for intended work time and intended break time?_  
- **API Endpoints** _Add more test cases to account for this._  

---

### **Final Notes**  
If there's any further tests you can think of or if you think I've misunderstood the intended functionality of certain fields, let me know!

### CODE REVIEW

**Review by aviendha-andrus — APPROVED** [review]
## PR Approval: Backend - Testing User Models
Reviewed the addition of appropriate tests for pomodoro model.

### General Feedback
1. `backend/apps/pomodoro/tests.py` is well done and looks clean.  
2. When I run `python3 manage.py test apps.submissions` or `coverage run manage.py test apps.submissions` all tests pass. 
2. Testing is an instrumental part of the process so its good that you added these tests, it looks like you covered all the main parts of testing the pomodoro app.

### Suggestions for Improvement
1. **Additional tests for edge cases**
* You might add tests for Invalid rarity choices, empty required fields, duration limits or other edge cases. 
2. **Coverage report**
* Try running `coverage run manage.py test apps.users` then `coverage report` to see what percentage of your code the tests cover. Right now for models.py you're at 87%.  

### Conclusion
This looks really good, great job!

File: backend/apps/pomodoro/tests.py [hunk 1]
```diff
@@ -1,3 +1,108 @@
...
+    
+    def test_str_representation(self):
+        pet = PetCatalog.objects.create(
+            name='Ruffles',
+            pet_type='Chocolate lab',
+            description='Just a baby',
+            rarity=PetCatalog.RARE,
+            drop_rate=0.3
+        )
+        self.assertEqual(str(pet), "Shadow (rare)")
```
Comment (ashley-arellano) [inline backend/apps/pomodoro/tests.py:67]: Please adjust this to: ```self.assertEqual(str(pet), "Ruffles (rare)")``` to pass the test

File: backend/apps/pomodoro/tests.py [hunk 2]
```diff
@@ -1,3 +1,108 @@
...
+
+    def test_invalid_drop_rate(self):
+        with self.assertRaises(IntegrityError):
```
Comment (ashley-arellano) [inline backend/apps/pomodoro/tests.py:70]: Please replace this code with: 
```
        pet = PetCatalog(
            name='InvalidPet',
            pet_type='Unknown',
            description='Invalid',
            rarity=PetCatalog.COMMON,
            drop_rate=1.5  # Invalid value
        )
        with self.assertRaises(ValidationError):
            pet.full_clean()  # This will trigger the MinValueValidator and MaxValueValidator
```

File: backend/apps/pomodoro/tests.py [hunk 3]
```diff
@@ -1,3 +1,108 @@
 from django.test import TestCase
```
Comment (ashley-arellano) [inline backend/apps/pomodoro/tests.py:2]: To begin, in your `apps/pomodoro/models.py` please add the following changes, then makemigrations + migrate:
Add this library: ```from django.core.validators import MinValueValidator, MaxValueValidator```
Adjust your `duration` field in `PomodoroSession` model: ```    duration = models.IntegerField(default=25, validators=[MinValueValidator(1)])```
Adjust your `drop_rate` field for `PetCatalog` model to: ```    drop_rate = models.FloatField(validators=[MinValueValidator(0.0), MaxValueValidator(1.0)])```
Please add to your  `class Meta` in `PetCollections` model: ```        unique_together = ('student', 'pet_catalog')  # Prevent duplicate pet collections```

**Review by ashley-arellano — CHANGES_REQUESTED** [review]
### 🔍 Peer Review **_TEMPLATE_**: Great Job on This PR!

Hey Allison 👋 — just finished reviewing your PR titled **"# Added Pomodoro and Gamification Tests"**. Here’s my feedback:

----------

## ✅ What Looks Great

-   Love the thorough test coverage (100% for `pomodoro/models.py`).
    
-   Your test structure is well-organized, covering various edge cases effectively.
    
-   The PR description is detailed and well-documented, making it easy to follow along.
    

----------

## 💡 Suggestions for Improvement (Fixing Failing Tests)

> _Here are the test issues I encountered and how they can be fixed._

1️⃣ **`test_invalid_drop_rate` (PetCatalogModelTest)**

-   **Issue:** The test expects an `IntegrityError`, but Django does not enforce constraints on `FloatField` at the database level in SQLite.
    
-   **Fix:** Use `full_clean()` to trigger Django’s `MinValueValidator` and `MaxValueValidator` (more details in the comments)

2️⃣ **`test_negative_duration` (PomodoroSessionModelTest)**

-   **Issue:** `IntegrityError` is not raised because SQLite does not enforce integer constraints.
    
-   **Fix:** Add `MinValueValidator(1)` in the model and use `full_clean()` in the test (more details in the comments)
    

3️⃣ **`test_duplicate_pet_collection` (PetCollectionModelTest)**

-   **Issue:** No uniqueness constraint is enforced for `(student, pet_catalog)`.
    
-   **Fix:** Add `unique_together = ('student', 'pet_catalog')` in `Meta` of `PetCollection` (more details in the comments)
    

4️⃣ **`test_create_pomodoro_session` (PomodoroSessionModelTest)**

-   **Issue:** The test expects `end_time` to be `None`, but your `save()` method sets it automatically.
    
-   **Fix:** Create the session with `is_completed=False` to prevent `end_time` from being set (more details in the comments)
    

----------

## 🧪 Tested This Feature

I followed the test instructions:

-   ✅ Ran `coverage run manage.py test apps.pomodoro` to check test results.
    
-   ✅ Verified the `coverage report` showing 100% test coverage for `models.py`.
    
-   ✅ Debugged and identified issues in failing tests.
    

----------

## 🔄 Next Steps

-   Implement the suggested test fixes and re-run the tests.
    
-   Let me know once the updates are made, and I’ll re-review before merging.
    

----------

### 🚀 Final Thoughts

Great effort on ensuring full test coverage and writing well-structured tests! A few minor tweaks should get everything passing. Good job : D

**Review by aviendha-andrus — APPROVED** [review]
Hi! Sorry, it looks like last time I was reviewing the wrong branch, I made sure to check the right branch this time, every test passes!

**Review by ashley-arellano — APPROVED** [review]
Thanks for adding the changes I requested! I'll approve this now : D
